File: tools/frida/dump_oura_models.py
#!/usr/bin/env python3
"""Drive the on-device Oura model decryption dump (dump_oura_models.js).

The models decrypt with a server-delivered key cached on a logged-in device
(docs/algorithms/sleepnet.md), so they cannot be decrypted from the APK alone.
This attaches Frida to the running Oura app, hooks its own AES-GCM decryption,
and keeps every plaintext that is a TorchScript container. You then navigate the
app so it loads the models you want; each decrypt lands in
/data/local/tmp/oura_models_dump/ on the device.

Nothing here prints or stores key material — only the decrypted model bytes,
which are the same artifacts docs/model-runners.md expects in notes/models/.

Prereqs: frida-server (matching frida major version, device arch) running as
root on the device; `pip install frida-tools` on the host.

    python tools/frida/dump_oura_models.py            # attach + hook, Ctrl-C to stop
    python tools/frida/dump_oura_models.py --pull      # after: match + pull to notes/models/
"""
from __future__ import annotations

import argparse
import subprocess
import sys
import time
from pathlib import Path

PKG = "com.ouraring.oura"
DEVICE_OUT = "/data/local/tmp/oura_models_dump"
HERE = Path(__file__).resolve().parent
SCRIPT = HERE / "dump_oura_models.js"
REPO = HERE.parent.parent
MODELS_OUT = REPO / "notes" / "models"


def sh(cmd: list[str], **kw) -> subprocess.CompletedProcess:
    return subprocess.run(cmd, text=True, capture_output=True, **kw)


def su(cmd: str) -> str:
    return sh(["adb", "shell", "su", "-c", cmd]).stdout


def _oura_pid() -> int | None:
    out = sh(["adb", "shell", "pidof", PKG]).stdout.strip()
    return int(out.split()[0]) if out.split() else None


def hook() -> int:
    import frida

    device = frida.get_usb_device(timeout=15)
    # Attach by pid straight from `pidof`: enumerate_applications() spawns a helper in
    # every installed app to read its label and times out badly under load. The agent
    # handshake can also close once while the app is busy (post-boot, mid-render), so
    # retry a few times before giving up.
    pid = _oura_pid()
    if pid is None:
        print(f"{PKG} is not running — launch the Oura app first, then re-run.", file=sys.stderr)
        return 2
    session = None
    for attempt in range(1, 6):
        try:
            session = device.attach(pid)
            break
        except frida.TransportError as e:
            print(f"attach {attempt}/5 failed ({e}); the app is busy — retrying", file=sys.stderr)
            time.sleep(3)
            pid = _oura_pid() or pid
    if session is None:
        print("could not attach — is frida-server running and its version matched?", file=sys.stderr)
        return 3
    script = session.create_script(SCRIPT.read_text())
    script.on("message", lambda msg, data: print("  frida:", msg.get("payload", msg)))
    script.load()
    print(
        "Hooks installed. In the Oura app, open the screens whose models you want:\n"
        "  Sleep  -> sleepnet_moonstone / sleepstaging\n"
        "  Activity/Workouts -> automatic_activity_detection, steps_motion_decoder\n"
        "  Cardiovascular age -> cva\n"
        "  Symptom Radar / illness -> illness_detection\n"
        "Each decrypt is written on the device. Press Ctrl-C here when done, then run "
        "with --pull.",
    )
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nDetaching.")
    session.detach()
    return 0


def pull() -> int:
    """Match each dumped plaintext to a .enc asset by size, then pull and name it."""
    listing = su(f"ls -1 {DEVICE_OUT}/*.pt 2>/dev/null")
    dumped = [line.strip() for line in listing.splitlines() if line.strip()]
    if not dumped:
        print(f"No dumps found in {DEVICE_OUT} — did the app load any models?", file=sys.stderr)
        return 1

    # Size of each dump on device.
    sizes: dict[str, int] = {}
    for path in dumped:
        out = su(f"stat -c %s {path}").strip()
        if out.isdigit():
            sizes[path] = int(out)

    # The .enc assets are 12 (IV) + 16 (GCM tag) bytes larger than their plaintext.
    enc_index = _enc_asset_sizes()
    MODELS_OUT.mkdir(parents=True, exist_ok=True)
    pulled = 0
    for path, size in sizes.items():
        name = enc_index.get(size + 28)
        local = MODELS_OUT / (name if name else f"unknown-{size}.pt")
        if sh(["adb", "exec-out", "su", "-c", f"cat {path}"], **{}).returncode != 0:
            continue
        with local.open("wb") as fh:
            proc = subprocess.run(["adb", "exec-out", "su", "-c", f"cat {path}"], stdout=fh)
        if proc.returncode == 0:
            tag = name or f"UNMATCHED ({size} bytes)"
            print(f"  pulled {tag}")
            pulled += 1
    print(f"{pulled} model(s) written to {MODELS_OUT}")
    print("Next: python tools/export_mobile.py   (produces the .ptl for Android/iOS)")
    return 0


def _enc_asset_sizes() -> dict[int, str]:
    """Map .enc asset size -> plaintext .pt filename, from the split models APK."""
    apk = su(
        "pm path com.ouraring.oura | sed 's/package://' | tr -d '\\r'"
    )
    paths = [p.strip() for p in apk.splitlines() if "oura_models" in p]
    index: dict[int, str] = {}
    for apk_path in paths:
        out = sh(["adb", "shell", f"unzip -l {apk_path}"]).stdout
        for line in out.splitlines():
            parts = line.split()
            if len(parts) >= 4 and parts[-1].startswith("assets/") and parts[-1].endswith(".pt.enc"):
                size = int(parts[0])
                name = Path(parts[-1]).name[: -len(".enc")]  # strip .enc -> <name>.pt
                index[size] = name
    return index


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--pull", action="store_true", help="Match + pull dumps to notes/models/")
    args = ap.parse_args()
    return pull() if args.pull else hook()


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/frida/dump_oura_models.js
/*
 * Dump Oura's decrypted on-device PyTorch models by hooking the app's OWN decryption
 * at runtime. The models ship as assets/<name>.pt.enc in split_oura_models.apk, encrypted
 * AES-256-GCM with a SERVER-DELIVERED key that is only cached (Keystore-wrapped) on a
 * logged-in device — see docs/algorithms/sleepnet.md. So they cannot be decrypted offline
 * from the APK, but the app itself decrypts them on demand, and this rides along.
 *
 * Rather than chase Oura's class names (which move between app versions), this hooks
 * javax.crypto.Cipher.doFinal and keeps any plaintext that is a TorchScript container —
 * a Zip (`PK\x03\x04`) or the legacy pickle magic. That is version-independent.
 *
 * Files are written on the DEVICE to /data/local/tmp/oura_models_dump/<sha256>.pt; the
 * driver (dump_oura_models.py) names them by matching size+hash against the .enc assets
 * and pulls them off. Frida is loud on stdout but never prints key material.
 *
 * Usage (via the driver): python tools/frida/dump_oura_models.py
 */
'use strict';

var OUT_DIR = '/data/local/tmp/oura_models_dump';

function isTorchScript(bytes) {
  if (!bytes || bytes.length < 8) return false;
  // Zip local file header "PK\x03\x04" — modern TorchScript / .ptl containers.
  if (bytes[0] === 0x50 && bytes[1] === 0x4b && bytes[2] === 0x03 && bytes[3] === 0x04) return true;
  // Legacy torch pickle: 0x80 0x02 (pickle proto 2) then 0x8a or '(' — be permissive.
  if (bytes[0] === 0x80 && (bytes[1] === 0x02 || bytes[1] === 0x03)) return true;
  return false;
}

Java.perform(function () {
  var Cipher = Java.use('javax.crypto.Cipher');
  var File = Java.use('java.io.File');
  var FOS = Java.use('java.io.FileOutputStream');
  var MessageDigest = Java.use('java.security.MessageDigest');

  try { File.$new(OUT_DIR).mkdirs(); } catch (e) {}

  var seen = {};
  var count = 0;

  function persist(bytes) {
    var md = MessageDigest.getInstance('SHA-256');
    var digest = md.digest(bytes);
    var hex = '';
    for (var i = 0; i < digest.length; i++) {
      hex += ('0' + (digest[i] & 0xff).toString(16)).slice(-2);
    }
    if (seen[hex]) return;
    seen[hex] = true;
    var path = OUT_DIR + '/' + hex + '.pt';
    var fos = FOS.$new(path);
    fos.write(bytes);
    fos.close();
    count++;
    console.log('[dump] wrote ' + bytes.length + ' bytes -> ' + hex + '.pt  (total ' + count + ')');
  }

  // Two doFinal overloads produce a fresh byte[]; those are the ones that hand back the
  // whole decrypted model. The into-buffer variants are used for streaming and ignored.
  ['doFinal', 'doFinal'].forEach(function () {});

  Cipher.doFinal.overload('[B').implementation = function (input) {
    var out = this.doFinal(input);
    try { if (isTorchScript(out)) persist(out); } catch (e) { console.log('[dump] err ' + e); }
    return out;
  };

  Cipher.doFinal.overload('[B', 'int', 'int').implementation = function (input, off, len) {
    var out = this.doFinal(input, off, len);
    try { if (isTorchScript(out)) persist(out); } catch (e) { console.log('[dump] err ' + e); }
    return out;
  };

  Cipher.doFinal.overload().implementation = function () {
    var out = this.doFinal();
    try { if (isTorchScript(out)) persist(out); } catch (e) { console.log('[dump] err ' + e); }
    return out;
  };

  // The docs describe PytorchModelFactory.createTemporaryFile: decrypt -> temp file ->
  // torch load -> delete. Catch that path too, class-name-independently, by copying any
  // temp file that is a TorchScript zip the instant before the app deletes it. This is
  // the backstop for a decrypt that streams (so Cipher.doFinal never sees whole bytes).
  var FIS = Java.use('java.io.FileInputStream');
  function persistFile(f) {
    try {
      if (!f.exists() || f.length() < 64) return;
      var head = Java.array('byte', [0, 0, 0, 0]);
      var fis = FIS.$new(f);
      fis.read(head); fis.close();
      if (!isTorchScript(head)) return;
      var all = Java.use('java.nio.file.Files').readAllBytes(f.toPath());
      persist(all);
    } catch (e) {}
  }
  File.delete.implementation = function () {
    try { persistFile(this); } catch (e) {}
    return this.delete();
  };

  console.log('[dump] hooks installed (Cipher.doFinal + temp-file preserve) — open Sleep/Activity');
});
